File: C#/myLeetCode/String/459. Repeated Substring Pattern.cs
// 459. Repeated Substring Pattern
// https://leetcode.com/problems/repeated-substring-pattern/

using System.Text;

public class Solution {
    public bool RepeatedSubstringPattern(string s) {
        int n = s.Length;
        for (int len = 1; len <= n / 2; len++) {
            if (n % len == 0) {  // Ensure the length is a factor of n
                int numRepeats = n / len;
                string pattern = s.Substring(0, len);
                StringBuilder concatenated = new StringBuilder();
                for (int i = 0; i < numRepeats; i++) {
                    concatenated.Append(pattern);
                }
                if (concatenated.ToString() == s) {
                    return true;
                }
            }
        }
        return false;
    }
}

public class Program {
    public static void Main() {
        Solution solution = new Solution();

        string input = "abab";
        bool result = solution.RepeatedSubstringPattern(input);

        Console.WriteLine($"Does '{input}' have a repeated substring pattern? {result}");
    }
}

File: C++/myLeetCode/String/459. Repeated Substring Pattern.cpp
// 459. Repeated Substring Pattern
// https://leetcode.com/problems/repeated-substring-pattern/

#include <iostream>

bool repeatedSubstringPattern(std::string s) {
    return (s + s).substr(1, 2 * s.size() - 2).find(s) != std::string::npos;
}

int main() {
    std::string str = "ababa";
    
    bool result = repeatedSubstringPattern(str);
    
    std::cout << "Is repeated substring pattern: " << std::boolalpha << result << std::endl;
    
    return 0;
}

// bool repeatedSubstringPattern(std::string str) {
//     int len = str.length();
    
//     for (int i = 1; i <= len / 2; i++) {
//         if (len % i == 0) {
//             std::string substring = str.substr(0, i);
//             std::string repeated = "";
            
//             for (int j = 0; j < len / i; j++) {
//                 repeated += substring;
//             }
            
//             if (repeated == str) {
//                 return true;
//             }
//         }
//     }
    
//     return false;
// }

